### DISCUSSION TITLE
Getting error while calling session.remove method on @app.teardown_appcontext

### DISCUSSION THREAD

**Discussion by bgshanker** [sqlalchemy/sqlalchemy#11294] [Usage Questions]
I am using render_template to render html page, have multiple tabs, loading each tab connects to database.
While navigating between tabs, i am getting following error while calling db.session.remove() method on app teardown_appcontext:

(cx_Oracle.DatabaseError) DPI-1080: connection was closed by ORA-3135

Any idea why this error is occurring? Do we really need to remove session on app teardown_appcontext?

**Comment by CaselIT** [comment]
Hi,

The description is not really helpful. Generally in a web request you would open a session at the start of the request and close it at the end of the request. I'm not really sure what `teardown_appcontext ` is here.

Does this happen after a long idle time or are timeout set on oracle?

**Reply by bgshanker** [reply]
We have flask returning html. There are multiple tabs in this page, clicking on each tab calls flask endpoint each connecting to database results in rendering a html page. At @app.teardown_appcontext, db session.remove() is called. The issue is not consistent, navigating between tabs sometimes results in connection already closed error as specified above. 

It looks like when flask endpoint is called, session is not active - not sure if it is because of the remove() method that is called in @app.teardown_appcontext or because of something else.

**Reply by CaselIT** [reply]
Seems like `@app.teardown_appcontext` is the right place to call `session.remove()`, but it's still unclear what you are doing.

Is session a scoped session? is there a timeout on the oracle side? have you tried using 
https://docs.sqlalchemy.org/en/20/core/engines.html#sqlalchemy.create_engine.params.pool_recycle or 
https://docs.sqlalchemy.org/en/20/core/engines.html#sqlalchemy.create_engine.params.pool_pre_ping ?

**Reply by bgshanker** [reply]
I hope the session is by default scoped_session. When i navigate between tabs, i am getting connection closed exception, not sure if it is because of sesion.remove() called in app.teardown_appcontext. Or because of pool_pre_ping, it is trying to close the connection. Is there a way to force reconnect if connection is closed?

This is my sqlalchemy db connection config:

SQLALCHEMY_ENGINE_OPTIONS = {
            'max_identifier_length' : 128,
            'pool_size' : 10,
            'pool_pre_ping' : True,
            'pool_use_lifo' : True,
            'pool_recycle' : -1
}
